### ISSUE TITLE
ENH: Type `pipe` with `ParamSpec`

### ISSUE THREAD

**Issue by paw-lu** [pandas-dev/pandas#56359]
### Feature Type

- [x] Adding new functionality to pandas

- [ ] Changing existing functionality in pandas

- [ ] Removing existing functionality in pandas

### Problem Description

Right now the `*args` and `**kwargs` in `pandas.DataFrame.pipe` are not typed, removing any type checking benefit on the parameters if a typed function is used via `pipe`.

https://github.com/pandas-dev/pandas/blob/2a953cf80b77e4348bf50ed724f8abc0d814d9dd/pandas/core/generic.py#L6039-L6044

### Feature Description

I'd be happy to contribute this if this seems acceptable.

I have a basic implementation here of what I'd want to apply to the [real method](https://github.com/pandas-dev/pandas/blob/v2.1.3/pandas/core/generic.py#L6037-L6142). I'm guessing this would go in [pandas-stubs](https://github.com/pandas-dev/pandas-stubs) as well?

One con here is that the tuple version of the `func` parameter does not use `ParamSpec`, since `Concatenate` requires that the additional type comes first. So users using a tuple for `func` wouldn't benefit from the type checking. Unfortunately, don't really know a way around this right now.

```python
from __future__ import annotations
import dataclasses
import typing
from typing import Callable, Concatenate

P = typing.ParamSpec("P")
T = typing.TypeVar("T")


@dataclasses.dataclass
class BabyDataFrame:
    data: int

    def pipe(
        self,
        func: Callable[Concatenate[BabyDataFrame, P], T] | tuple[Callable[..., T], str],
        *args: P.args,
        **kwargs: P.kwargs,
    ) -> T:
        if isinstance(func, tuple):
            return func[0](*args, **kwargs, **{func[1]: self})

        else:
            return func(self, *args, **kwargs)


def foo(baby_data_frame: BabyDataFrame, x: int) -> int:
    return baby_data_frame.data + x


def bar(x: int, baby_data_frame: BabyDataFrame) -> int:
    return baby_data_frame.data + x


print(BabyDataFrame(2).pipe(foo, 1))
print(BabyDataFrame(2).pipe((bar, "baby_data_frame"), 2))
```

### Alternative Solutions

Not sure of an alternative solution here besides just implementing in [pandas-stubs](https://github.com/pandas-dev/pandas-stubs). Figured I might as well do both.

### Additional Context

[`pipe` in pandas-stubs](https://github.com/pandas-dev/pandas-stubs/blob/2aadf0ee9cfc3ed3436f33a313ead9588216574c/pandas-stubs/core/generic.pyi#L355-L357).

**Comment by twoertwein** [comment]
I think it would be best first to have a pandas-stubs PR (with typing tests) and after that is merged and works well a pandas PR.

@Dr-Irv

**Comment by Dr-Irv** [comment]
> I think it would be best first to have a pandas-stubs PR (with typing tests) and after that is merged and works well a pandas PR.
> 
> @Dr-Irv

I agree.  The goal is to improve type checking for users, which is why we have `pandas-stubs` .

Also, see https://github.com/pandas-dev/pandas-stubs/issues/738

**Comment by paw-lu** [comment]
Alright, sounds good. I'll continue this in pandas-dev/pandas-stubs, then follow up here if things go well there. Thanks!

**Comment by paw-lu** [comment]
This was implemented and merged in

- https://github.com/pandas-dev/pandas-stubs/pull/823

Would like to port this over into [pandas-dev](https://github.com/pandas-dev/pandas), but I think the minimum python version here is 3.9. And I don't think `typing_extensions` is an existing dependency. This means that we can't use `ParamSpec` here right (which was introduced in 3.10)? Which means this can't be implemented here until it moves to 3.10.
